src/utils.c

```diff
@@ -826,3 +826,192 @@ gchar *r_bytes_unref_to_string(GBytes **bytes)
 
 	return g_strndup(data, size);
 }
+
+gboolean semver_parse(const gchar *version_string, guint64 version_core[3], gchar **pre_release, gchar **build, GError **error)
+{
+	g_autofree gchar *version_copy = g_strdup(version_string);
+	gchar *build_pos = NULL;
+	gchar *pre_release_pos = NULL;
+	g_auto(GStrv) version_parts = NULL;
+	GError *ierror = NULL;
+	int i = 0;
+
+	g_return_val_if_fail(version_string, FALSE);
+	g_return_val_if_fail(version_core, FALSE);
+	g_return_val_if_fail(pre_release == NULL || *pre_release == NULL, FALSE);
+	g_return_val_if_fail(build == NULL || *build == NULL, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	/* semantic versions BNF:
+	 * <valid semver> ::= <version core>
+	 * | <version core> "-" <pre-release>
+	 * | <version core> "+" <build>
+	 * | <version core> "-" <pre-release> "+" <build>
+	 *
+	 * detect first the '+build', separate it out by null terminating the
+	 * version_copy at it's start position; repeat for the '-pre-release'; and
+	 * then parse the remaining 'version core'
+	 */
+
+	build_pos = g_strrstr(version_copy, "+");
+	if (build_pos != NULL) {
+		if (build != NULL) {
+			*build = g_strdup(build_pos + 1);
```

Comment:
It would possibly simplify the flow here and later if we'd save the return values internally in an autopointer gchar and steal the pointer at the end with returning TRUE. But it works this way, too.

Reply:
I'd like it keep it this way for now.